File: Algorithm Templates/java/H02_SearchAlgorithms/BinarySearch.java
package H02_SearchAlgorithms;

/**
 * 二分查找
 */
public class BinarySearch {
    static int target;

    /**
     * 查找左边界，即第一个满足条件的元素下标 (lower_bound)
     */
    public static int binarySearchL(int l, int r) {
        while (l < r) {
            int mid = l + r >> 1;    // 计算中间值
            if (ge(mid, target)) {
                r = mid;            // 如果中间的值符合条件，则继续在左边查找
            } else {
                l = mid + 1;        // 否则在右边查找
            }
        }
        return l;    // 返回左边界
    }

    /**
     * 查找右边界，即最后一个满足条件的元素下标 (upper_bound的前驱)
     */
    public static int binarySearchR(int l, int r) {
        while (l < r) {
            int mid = l + r + 1 >> 1;    // 计算中间值，向右偏移
            if (le(mid, target)) {
                l = mid;                // 如果中间的值符合条件（≤），则继续在右边查找
            } else {
                r = mid - 1;            // 否则在左边查找
            }
        }
        return r;    // 返回右边界
    }

    /**
     * 浮点数二分
     */
    public static double binarySearchF(double l, double r) {
        final double eps = 1e-8;        // 精度，视题目而定
        while (r - l > eps) {
            double mid = (l + r) / 2;    // 计算中间值
            if (ge(mid, target)) {
                r = mid;        // 目标在左边，更新右边界
            } else {
                l = mid;        // 否则更新左边界
            }
        }
        return l;    // 返回左边界，即为目标值的估计
    }

    private static boolean ge(int mid, int target) {
        // 自定义比较条件 ...
        return false;
    }

    private static boolean ge(double mid, double target) {
        // 自定义比较条件 ...
        return false;
    }

    private static boolean le(int mid, int target) {
        // 自定义比较条件 ...
        return false;
    }
}

File: Algorithm_Templates/go/basic_algorithms/binary_search/binary_search.go
package binary_search

var (
	target1 int
	target2 float64
)

// BinarySearchL 查找左边界，即第一个满足条件的元素下标 (lower_bound)
func BinarySearchL(l, r int) int {
	for l < r {
		mid := (l + r) >> 1
		if ge(mid, target1) {
			r = mid // 如果中点值符合条件，则继续在左边查找
		} else {
			l = mid + 1
		}
	}
	return l
}

// BinarySearchR 查找右边界，即最后一个满足条件的元素下标 (upper_bound的前驱)
func BinarySearchR(l, r int) int {
	for l < r {
		mid := (l + r + 1) >> 1
		if le(mid, target1) {
			l = mid // 如果中点值符合条件（≤），则继续在右边查找
		} else {
			r = mid - 1
		}
	}
	return r
}

func BinarySearchF(l, r float64) float64 {
	for l < r {
		mid := (l + r) / 2
		if ge2(mid, target2) {
			r = mid // 目标在左边，更新右边界
		} else {
			l = mid
		}
	}
	return l // 返回左边界，即为目标值的估计
}

func ge(mid, target int) bool {
	// ...
	return mid >= target
}

func ge2(mid float64, target float64) bool {
	// ...
	return mid-target >= 0
}

func le(mid, target int) bool {
	// ...
	return mid <= target
}
